src/status_im/hardwallet/common.cljs

```diff
@@ -0,0 +1,431 @@
+            (restore-on-card-connected)
+            (restore-on-card-read)
+            (bottom-sheet/hide-bottom-sheet)))
+
+(fx/defn clear-pin
+  [{:keys [db] :as cofx}]
+  (fx/merge cofx
+            {:db (update-in db
+                            [:hardwallet :pin]
+                            merge
+                            {:status       nil
+                             :login        (get-in db [:hardwallet :pin :original])
+                             :export-key   []
+                             :sign         []
+                             :puk          []
+                             :current      []
+                             :original     []
+                             :confirmation []
+                             :error-label  nil})}))
+
+(fx/defn cancel-sheet-confirm
+  {:events [::cancel-sheet-confirm
+            :hardwallet/back-button-pressed]}
+  [cofx]
+  (fx/merge cofx
+            (hide-pair-sheet)
+            (clear-pin)))
+
+(fx/defn cancel-sheet
+  {:events [::cancel-sheet]}
+  [_]
+  {:ui/show-confirmation {:title               (i18n/label :t/keycard-cancel-setup-title)
+                          :content             (i18n/label :t/keycard-cancel-setup-text)
+                          :confirm-button-text (i18n/label :t/yes)
+                          :cancel-button-text  (i18n/label :t/no)
+                          :on-accept           #(re-frame/dispatch [::cancel-sheet-confirm])
+                          :on-cancel           #()}})
+
+(fx/defn on-add-listener-to-hardware-back-button
+  "Adds listener to hardware back button on Android.
+  During keycard setup we show user a warning that setup will be cancelled
+  when back button pressed. This prevents user from going back during setup
+  flow as some of the actions changing keycard step could not be repeated."
+  {:events [:hardwallet/add-listener-to-hardware-back-button]}
+  [{:keys [db]} listener]
+  {:db (assoc-in db [:hardwallet :back-button-listener] listener)})
+
+(fx/defn show-wrong-keycard-alert
+  [_ card-connected?]
+  (when card-connected?
+    {:utils/show-popup {:title   (i18n/label :t/wrong-card)
+                        :content (i18n/label :t/wrong-card-text)}}))
+
+(fx/defn unauthorized-operation
+  [cofx]
+  (fx/merge cofx
+            {:utils/show-popup {:title   ""
+                                :content (i18n/label :t/keycard-unauthorized-operation)}}
+            (clear-on-card-connected)
+            (navigation/navigate-to-cofx :keycard-settings nil)))
+
+(fx/defn navigate-to-enter-pin-screen
+  {:events [:hardwallet/navigate-to-enter-pin-screen]}
+  [{:keys [db] :as cofx}]
+  (let [key-uid               (get-in db [:hardwallet :application-info :key-uid])
+        multiaccount-key-uid  (get-in db [:multiaccount :key-uid])
+        keycard-multiaccount? (boolean (get-in db [:multiaccount :keycard-pairing]))]
+    ;; TODO: WHY we check key-uid? If last oepration was with wrong card
```

Comment:
is that a new comment? if so can you please elaborate?

Reply:
Yes, If for example, a user wants to sign with the keycard and uses a wrong card, then cancel, then they go to change pin instead of seeing the process to change pin, unauthorized error appear and they can't do anything. This happens because the latest read of the card is stored in DB and we check based on it. But the user's intention is to use a different card, the correct one. IMO ideally this information should not be in DB and only read and passed as parameters to functions that needs it. 
Right now we have a lot of information related to keycard in DB and events work based on that info. This makes it harder to follow how state is passed through the flow. There are more errors like that, like the pin is not cleared in some places, and on some flows redirect is not correct.